src/lazy/expanded/mod.rs

```diff
@@ -1117,6 +1117,14 @@ impl<'top, Encoding: Decoder> LazyExpandedValue<'top, Encoding> {
         }
         None
     }
+
+    #[cfg(feature = "source-location")]
+    pub fn location(&self) -> Option<(usize, usize)> {
+        if let ExpandedValueSource::ValueLiteral(value) = &self.source {
+            return Some(value.location());
+        }
+        None
+    }
```

Comment:
What would you think of something like this? We can have the function always be part of the public API (so that it's easier to turn on/off), but when the feature is not enabled, it always returns `None`.
```suggestion
    pub fn location(&self) -> Option<(usize, usize)> {
        #[cfg(feature = "source-location")]
        if let ExpandedValueSource::ValueLiteral(value) = &self.source {
            return Some(value.location());
        }
        None
    }
```

Reply:
I meant to hide the location information when the flag is not enabled. I think having it with `None` will work in the case though.

Reply:
I think this is particular to the `ion-schema-rust` usecase and I think the point of having the feature flag is that the API is only available under the feature flag. Also, we tried several benchmarks previously to decide on whether to keep `Option` vs just hide the properties under the feature flag and found that the feature flag was closest to zero-cost solution, which is why we decided to hide it under the feature flag in the last PR. 
I am just going by those results and going with feature-flag everywhere. But that being said, I can modify this to be an `Option` if that is something that is needed within `ion-schema-rust`.

Reply:
> I think this is particular to the ion-schema-rust usecase

Even if it is particular to `ion-schema-rust`, it would still make `ion-schema-rust` a nicer library for users if `ion-schema-rust` can automatically work both with and without the `source-location` feature flag.

> I think the point of having the feature flag is that the API is only available under the feature flag

That's not quite correct. The goal isn't to gate the feature—the reason we have a feature flag is to mitigate the performance impact caused by the feature.

Maybe I didn't explain my suggestion well enough. I am suggesting that we _keep_ the feature flag. When the feature flag is enabled, it looks pretty much like it does in this PR. However, when the feature flag is _not_ enabled:
* the internal bookkeeping code is not included in the compilation unit
  * as in this (and the prior) PR
* any internal fields used for tracking the locations can still exist, but they are switched to zero-sized types
  * I suggested this elsewhere in this PR; that's for readability, but it's a two-way decision, so I won't fight for it if you disagree
* the public API remains unchanged—that is to say `pub fn location(&self)` still exists, but it has a trivial implementation that always returns `None`.
  * This should not have any performance impact because it's not adding to the memory footprint of any of the types. It's also not adding additional computation anywhere—it's returning a constant, so even the function call is easily inlined by the compiler.

Does it make more sense if I re-phrase the suggested change like this?
```suggestion
    #[cfg(feature = "source-location")]
    pub fn location(&self) -> Option<(usize, usize)> {
        if let ExpandedValueSource::ValueLiteral(value) = &self.source {
            return Some(value.location());
        }
        None
    }
    
    #[cfg(not(feature = "source-location"))]
    pub fn location(&self) -> Option<(usize, usize)> { None }
```

Reply:
Yes, that makes sense. I previously did not get the part where you said this was for when the feature isn't enabled as you mentioned something about the performance measurement. Bu this makes sense and I am happy to add this change. 

Were you referring to your comment [here](https://github.com/amazon-ion/ion-rust/pull/922#pullrequestreview-2659472670) for the zero-sized types? I do think its useful for readability(Which is why I agree that should be changed) but I do not think its gonna change performance as it would still require the same two proprieties `prev_newline_offset` and `row` for storing location metadata.
Also, the reason I am saying this is because I am trying to to use the new `LazyElement` changes where I can count newlines based on `IonInput` whenever more iput is read into buffer instead of calculating that at `TextBuffer`. which is why I am only focusing on changes that could impact the current implementation performance so that I can compare it with the new suggested change once its ready.(which might likely have better performance)

Reply:
> Were you referring to your comment https://github.com/amazon-ion/ion-rust/pull/922#pullrequestreview-2659472670 for the zero-sized types?

Yes. I don't think it's going to impact performance, which is why it should be okay. I can explain in more detail:

Instead of passing around two fields for the `row` and `previous_newline_offset`, we can encapsulate that into a type such as `pub(crate) LocationTrackingState { row: usize, previous_newline_offset: usize }`. That reduces some of the noise in the code simply by going from two fields/arguments to one. However, the more interesting thing is that it allows us to use a trick involving _zero-sized types_.

Instead of conditionally compiling fields and parameters that store `LocationTrackingState` based on the feature flag, we change the _declaration_ of `LocationTrackingState` based on the feature flag:
```rust
#[cfg(feature="source-location")]
#[derive(Default, Copy)]
pub(crate) struct LocationTrackingState {
    row: usize,
    previous_newline_offset: usize
}

#[cfg(not(feature="source-location"))]
#[derive(Default, Copy)]
pub(crate) struct LocationTrackingState;
```

Now, we don't need to have conditionally compiled function arguments or struct fields anywhere that `LocationTrackingState` is used because when the feature flag is _not_ enabled, `LocationTrackingState` is a [zero-sized type](https://doc.rust-lang.org/nomicon/exotic-sizes.html#zero-sized-types-zsts), and "Rust largely understands that any operation that produces or stores a ZST can be reduced to a no-op", so the cost of having those fields should be nothing.

This also makes it easy to identify which code needs to have the feature flag on it—it's only code that needs to read or write the fields of `LocationTrackingState` (so, for example, the conditional code in that checkpoint function no longer needs to be conditional).

Reply:
Sorry, If I was not clear earlier I do understand that this makes sense. I was trying to explain why I am holding off on changes that do not affect performance.